dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java

```diff
@@ -0,0 +1,558 @@
+import io.etcd.jetcd.lease.LeaseGrantResponse;
+import io.etcd.jetcd.lease.LeaseRevokeResponse;
+import io.etcd.jetcd.lease.LeaseTimeToLiveResponse;
+import io.etcd.jetcd.options.DeleteOption;
+import io.etcd.jetcd.options.GetOption;
+import io.etcd.jetcd.options.LeaseOption;
+import io.etcd.jetcd.options.WatchOption;
+import io.etcd.jetcd.watch.WatchEvent;
+import io.etcd.jetcd.watch.WatchResponse;
+import io.netty.util.internal.StringUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Wrapper class around jetcd client to achieve utilities API to talk with ETCD.
+ */
+public class AlluxioEtcdClient implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEtcdClient.class);
+  private static final Lock INSTANCE_LOCK = new ReentrantLock();
+  @GuardedBy("INSTANCE_LOCK")
+  private static final AtomicReference<AlluxioEtcdClient> ALLUXIO_ETCD_CLIENT
+      = new AtomicReference<>();
+  private final AtomicBoolean mConnected = new AtomicBoolean(false);
+  private Client mClient;
+  public final ServiceDiscoveryRecipe mServiceDiscovery;
+  public String[] mEndpoints;
+  private final Closer mCloser = Closer.create();
+  // only watch for children change(add/remove) for given parent path
+  private final ConcurrentHashMap<String, Watch.Watcher> mRegisteredWatchers =
+      new ConcurrentHashMap<>();
+
+  /**
+   * CTOR for AlluxioEtcdClient.
+   * @param conf
+   */
+  public AlluxioEtcdClient(AlluxioConfiguration conf) {
+    String clusterName = conf.getString(PropertyKey.ALLUXIO_CLUSTER_NAME);
+    List<String> endpointsList = conf.getList(PropertyKey.ETCD_ENDPOINTS);
+    mEndpoints = endpointsList.toArray(new String[endpointsList.size()]);
+    mServiceDiscovery = new ServiceDiscoveryRecipe(this, clusterName);
+  }
+
+  /**
+   * Get the singleton instance of AlluxioEtcdClient.
+   * @param conf
+   * @return AlluxioEtcdClient
+   */
+  public static AlluxioEtcdClient getInstance(AlluxioConfiguration conf) {
+    if (ALLUXIO_ETCD_CLIENT.get() == null) {
+      try (LockResource lockResource = new LockResource(INSTANCE_LOCK)) {
+        if (ALLUXIO_ETCD_CLIENT.get() == null) {
+          ALLUXIO_ETCD_CLIENT.set(new AlluxioEtcdClient(conf));
+        }
+      }
+    }
+    return ALLUXIO_ETCD_CLIENT.get();
```

Comment:
So you prefer the readers to not hold a lock right? What about `volatile` instead of `AtomicRef`?

Reply:
Also for a non-trivial pattern as the lazy-init you are doing here, I think it's worth a little javadoc comments explaining the thread safety requirements and guarantees. Eg. https://github.com/Alluxio/alluxio/blob/982d0677a2920da0093e3661e24cd1ce27595e4e/dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java#L81

Reply:
done

Reply:
i actually dont think there's a necessity to set volatile as long as u r guraded by instantiation lock as only party who would reassign value to the variable is the instantiation.